README.md
# SolidusClient

Render your [Solidus](https://github.com/solidusjs/solidus) views in the browser! Just like Solidus, this library will:

 1. Fetch data from static and dynamic resources, and make it available in the context.
 2. Add and transform the context with custom preprocessors.
 3. Render [Handlebars](http://handlebarsjs.com/) templates, with partials and helpers. The [handlebars-helper](https://github.com/SparkartGroupInc/handlebars-helper) helpers are automatically made available.

Refer to the Solidus documentation to know more about those concepts.

# Usage

Rendering a template starts with the `render` function call, and ends with the `end` callback. The various options can be passed directly to the `render` method as a config object, or by chaining method calls on the returned object.

 - `resources` - Object of named urls or options. The urls can have dynamic segments, which will be replaced by the matching `params`. If an error occurs while fetching a resource, the template is not rendered, except if the resource's `optional` option is `true`.
  ```javascript
  var resources = {
    blogs: 'http://my-site.com/blogs?page={page}',
    news: {
      url: 'http://news-site.com/news',
      proxy: true
    }
  }
  };
  ```

  Available options:
   - `url` - Resource URL.
   - `query` - Object of query values to add to the resource URL.
   - `query_options` - Options passed to [`qs.stringify`](https://github.com/hapijs/qs#stringifying), the function used to serialize the `query` object. An additional option is available, `objectFormat: 'json'`, which will encode query values of type object into JSON strings.
   - `headers` - Object of HTTP headers to add to the resource request.
   - `auth` - Object with `user` and `pass` values for HTTP Basic authentication.
   - `with_credentials` - Whether to send cookies from the origin. Defaults to `false`.
   - `jsonp` - Whether to use jsonp instead of ajax for the request. Defaults to `false`.
   - `proxy` - Whether to fetch the resource through Solidus, instead of directly hitting the resource URL. If `true`, all other options are ignored. Defaults to `false`.
   - `solidus_api_route` - Route to the Solidus `resource.json` endpoint if `proxy` is `true`. Defaults to `/api/`.
   - `timeout` - Maximum time to wait for the resource, in milliseconds. Defaults to 20,000.
   - `optional` - If `true` and the resource cannot be fetched, the template is rendered anyway. Defaults to `false`. Error conditions:
     - Invalid URL.
     - A required `param` is missing.
     - HTTP error, like a network error.
     - Timeout is exceeded.
     - Returned status code is not in the 2xx range.
     - Returned data is not valid JSON.
     - Returned data has a root property named `status` with value `error`.
 - `params` - Object of named values. The values will be interpolated into the dynamic resource urls.
  ```javascript
  var params = {
    page: 123
  };
  ```

 - `required_params` - Array of required param names. If any param is missing, an error is returned.
 - `preprocessor` - Function that modifies the context. The preprocessor is run after the resources are fetched, but before the template is rendered. If an error occurs while running the preprocessor, the template is not rendered.
  ```javascript
  // Sync
  var preprocessor = function(context) {
    context.blogs_count = context.resources.blogs.length;
    return context;
  };
  ```

  ```javascript
  // Async
  var preprocessor = function(context, callback) {
    context.blogs_count = context.resources.blogs.length;
    if (context.resources.blog.some_condition) {
      solidus_client.getResource('http://www.my-site.com/something', null, function(err, data) {
        if (err) throw err;
        context.more_data = data;
        callback(context);
      });
    } else {
      callback(context);
    }
  };
  ```

  Return values:
   - `object` - Modified context.
   - `string` - URL to immediately redirect the browser to, instead of rendering the template.
   - `array` - Item 0 is ignored, item 1 is URL to immediately redirect the browser to, instead of rendering the template.
   - Any other value results in a preprocessor error.

 - `template` - Compiled [Handlebars template](http://handlebarsjs.com/).
  ```javascript
  var template = Handlebars.compile('{{blogs_count}} posts: <ul>{{#each resources.blogs}}<li>{{> blog}}</li>{{/each}}</ul>');
  ```

 - `template_options` - Object of [options to use](http://handlebarsjs.com/execution.html) when rendering the template: `data`, `helpers` and `partials`.
  ```javascript
  var template_options = {
    helpers: {
      uppercase: function(string) {
        return string.toUpperCase();
      }
    },
    partials: {
      blog: Handlebars.compile('{{uppercase name}}');
    }
  };
  ```

## Config Object Example

```javascript
var view = {
  resources: resources,
  params: params,
  preprocessor: preprocessor,
  template: template,
  template_options: template_options
};

var solidus_client = new SolidusClient();
solidus_client.render(view)
  .end(function(err, html) {
    console.log(html);
  });
```

## Methods Chain Example

```javascript
var solidus_client = new SolidusClient();
solidus_client.render(template)
  .params(params)
  .template_options(template_options)
  .get(resources)
  .then(preprocessor)
  .end(function(err, html) {
    console.log(html);
  });
```

## Combined Example

```javascript
var view = {
  resources: resources,
  preprocessor: preprocessor,
  template: template,
  template_options: template_options
};

var solidus_client = new SolidusClient();
solidus_client.render(view)
  .params(params)
  .end(function(err, html) {
    console.log(html);
  });
```

# Auth

Just like in Solidus, resource security settings are [configured globally](https://github.com/solidusjs/solidus#global-resource-configuration). The `resources_options` property of the `SolidusClient` instance is scanned whenever a resource is fetched, to find matching options. See the Resources documentation above for the available options. Example:

```javascript
var resources_options = {
  "http://proxy.storyteller.io/*": {
    headers: {
      "Api-Key": "0000aaaa-aa00-00aa-a00a-aaaa000000"
    }
  },
  "http://services.sparkart.net/*": {
    query: {
      key: "1111bbbb-bb11-11bb-b11b-bbbb111111"
    }
  }
};

// Two ways to assign the resources options
var solidus_client = new SolidusClient({resources_options: resources_options});
solidus_client.resources_options = resources_options;
```

# Building

```
$ npm run build
```

# Testing

## Node and automated browser tests

```
$ npm run test
```

## Node test

```
$ npm run node-test
```

## Automated browser test

```
$ npm run browser-test
```

## Manual browser test

```
$ npm run test-server
```
Then access [http://localhost:8081/test/browser/test.html](http://localhost:8081/test/browser/test.html) in your browser


CHANGELOG.md
## 1.4.1 (Jan 29, 2021)

 - Update extend, to prevent __proto__ attack [[a376c83](https://github.com/solidusjs/solidus-client/commit/a376c831087a1745c00977dedc274f65451d21da)]

## 1.4.0 (Nov 6, 2018)

 - Required params [[7de41bb](https://github.com/solidusjs/solidus-client/commit/7de41bb21c1c3731c0edf9d25655e589ce370e6d)]

## 1.3.5 (Apr 23, 2018)

 - Use BrowserStack real devices [[5d4113b](https://github.com/solidusjs/solidus-client/commit/5d4113b4edad97ae1968e8f98576383a8761f489)]
 - Update SuperAgent to 3.8.2 [[c8d2ab3](https://github.com/solidusjs/solidus-client/commit/c8d2ab36362b34397d6245a407823d59c20ac7bf)]

## 1.3.4 (Nov 28, 2017)

 - Update SuperAgent to 3.8.1 [[0aec006](https://github.com/solidusjs/solidus-client/commit/0aec006301bf18f311c02deab676b61b540f2104)]

## 1.3.3 (Dec 18, 2015)

 - Remove resource.key() [[a88aa1b](https://github.com/solidusjs/solidus-client/commit/a88aa1b80608f4b0444d0ab31461782fb4d681e8)]

## 1.3.2 (Nov 27, 2015)

 - Fixed breaking test [[1d82f1f](https://github.com/solidusjs/solidus-client/commit/1d82f1f14b7649fd9d2b96f3beb0d5e2e3385840)]

## 1.3.1 (Nov 25, 2015)

 - Check if WordPress revision terms are blank and replace them with original terms for previews [[ed1fc0c](https://github.com/solidusjs/solidus-client/commit/ed1fc0cabe8acd4b488096f0a80a3bb84bd67385)]

## 1.3.0 (Nov 10, 2015)

 - Add WordPress helper library with .getPostPreviewView function [[24cd1f6](https://github.com/solidusjs/solidus-client/commit/24cd1f664e9fc7dbe07b27ac09d7e8239d7f04ee)]

## 1.2.2 (Apr 16, 2015)

 - Fix CORS and JSONP on older browsers [[2ba8659](https://github.com/solidusjs/solidus-client/commit/2ba86591c8e6419a1b4a94a77780598f77562eb3)]

## 1.2.1 (Apr 15, 2015)

 - Fix JSONP requests for non-ready body element [[ec7c259](https://github.com/solidusjs/solidus-client/commit/ec7c25989754cf65791504ec8a956c347a8cd5f6)]
 - Tweak CORS support [[1a7086f](https://github.com/solidusjs/solidus-client/commit/1a7086f70d0075ecf568c116a95b191cbd35bae9)]

## 1.2.0 (Apr 13, 2015)

 - Add resource.key(), fix proxy url to include query part [[a8f72d5](https://github.com/solidusjs/solidus-client/commit/a8f72d5bc3b09fc6b46553dd525c872ff010285e)]

## 1.1.1 (Apr 9, 2015)

 - Validate format of redirect array returned by preprocessors [[e820d66](https://github.com/solidusjs/solidus-client/commit/e820d660dcd4c09c6e4a16e2d71ed50a63aa87bd)]

## 1.1.0 (Apr 8, 2015)

 - Query options [[00acb21](https://github.com/solidusjs/solidus-client/commit/00acb21a02a6ccd9824e5e6a914a92ff8f74c783)]

## 1.0.0 (Apr 1, 2015)

 - Initial release


test/test.js
require('./tests/resource')();
require('./tests/solidus-client')();
require('./tests/view')();
require('./tests/wordpress')();


test/view.js
var assert = require('assert');
var Handlebars = require('handlebars');
var nock = require('nock');

var SolidusClient = require('../index.js');

describe('View', function() {
  var solidus_client = new SolidusClient();
  var resources = {test: 'http://solidus.{a}'};
  var preprocessor = function(context) {
    context.test = context.resources.test.test;
    return context;
  };
  var template = Handlebars.compile('{{test}} {{custom}}');
  var template_options = {helpers: {custom: function() {return 'helped';}}};
  var params = {a: 'com'}

  beforeEach(function() {
    nock.disableNetConnect();
    nock('http://solidus.com').get('/').reply(200, '{"test": "success!"}');
  });

  afterEach(function() {
    nock.enableNetConnect();
  });

  describe('with view object', function() {
    it('with all options', function(done) {
      var view = {
        resources: resources,
        params: params,
        preprocessor: preprocessor,
        template: template,
        template_options: template_options
      };

      var solidus_client = new SolidusClient();
      solidus_client
        .render(view)
        .end(function(html) {
          assert.equal(html, 'success! helped');
          assert.deepEqual(solidus_client.context, {resources:{test:{test:'success!'}},test:'success!'});
          done();
        });
    });

    it('with only template', function(done) {
      var solidus_client = new SolidusClient();
      solidus_client
        .render({template: Handlebars.compile('works! {{resources.test}}')})
        .end(function(html) {
          assert.equal(html, 'works! ');
          assert.deepEqual(solidus_client.context, {resources:{}});
          done();
        });
    });
  });

  describe('with methods chain', function() {
    it('with all options', function(done) {
      var solidus_client = new SolidusClient();
      solidus_client
        .render(template)
        .params(params)
        .templateOptions(template_options)
        .get(resources)
        .then(preprocessor)
        .end(function(html) {
          assert.equal(html, 'success! helped');
          assert.deepEqual(solidus_client.context, {resources:{test:{test:'success!'}},test:'success!'});
          done();
        });
    });

    it('with only template', function(done) {
      var solidus_client = new SolidusClient();
      solidus_client
        .render(Handlebars.compile('works! {{resources.test}}'))
        .end(function(html) {
          assert.equal(html, 'works! ');
          assert.deepEqual(solidus_client.context, {resources:{}});
          done();
        });
    });
  });

  it('with async preprocessor', function(done) {
    var solidus_client = new SolidusClient();
    solidus_client
      .render(Handlebars.compile('works! {{test}}'))
      .then(function(context, callback) {
        solidus_client.getResources(resources, params, function(err, resources) {
          assert.ifError(err);
          context.test = resources.test.test;
          callback(context);
        });
      })
      .end(function(html) {
        assert.equal(html, 'works! success!');
        assert.deepEqual(solidus_client.context, {resources:{},test:'success!'});
        done();
      });
  });
});


test/config.js
var fs   = require('fs');
var zlib = require('zlib');
var url  = require('url');

// Safari doesn't like localhost, see https://www.browserstack.com/question/663
module.exports.port = 8081;
module.exports.host = 'http://lvh.me:' + module.exports.port;

module.exports.routes = function (req, res) {
  if (req.url.indexOf('/test/browser/') == 0) {
    fs.readFile('.' + req.url.replace(/\?.*$/, ''), function(err, data) {
      var content_type;
      if (/\.css$/.test(req.url)) content_type = 'text/css';
      else if (/\.js$/.test(req.url)) content_type = 'text/javascript';
      else content_type = 'text/html';
      res.writeHead(200, {'Content-Type': content_type});
      res.end(err || data);
    });
    return;
  }

  var success = function(data) {
    data || (data = {url: req.url.replace(/solidus_client_jsonp_callback_\d+/, 'solidus_client_jsonp_callback')});
    res.writeHead(200, {'Content-Type': 'application/json'});
    var callback = url.parse(req.url, true).query.callback;
    if (callback) {
      res.end(callback + '(' + JSON.stringify(data) + ');');
    } else {
      res.end(JSON.stringify(data));
    }
  };

  var not_found = function() {
    res.writeHead(404, {'Content-Type': 'application/json'});
    res.end(JSON.stringify({error: 'Not Found: ' + req.url}));
  };

  switch ((req.method + ' ' + req.url).replace(/solidus_client_jsonp_callback_\d+/, 'solidus_client_jsonp_callback')) {
  case 'GET /page':
  case 'GET /page?a=':
  case 'GET /page?a=1':
  case 'GET /page?a=1&b=2':
  case 'GET /page?a=1&b=3&c=4':
  case 'GET /page?a=%2C&b=%2C&c=%2C&d=%252C':
  case 'GET /page?a=1&b=2&c%5B0%5D=3&c%5B1%5D=4&d%5Be%5D=5&d%5Bf%5D%5B0%5D=6&d%5Bf%5D%5B1%5D=7':
  case 'GET /page?a=1&b=2&c%5B0%5D=3&c%5B1%5D=4&d=%7B%22e%22%3A5%2C%22f%22%3A%5B6%2C7%5D%7D':
  case 'GET /api/resource.json?url=http%3A%2F%2Fsolidus.com%3Fa%3D1':
  case 'GET /custom/route/resource.json?url=http%3A%2F%2Fsolidus.com':
  case 'GET /page?a=1&b=2&callback=solidus_client_jsonp_callback':
  case 'GET /with-post-object?a=1&b=2&callback=solidus_client_jsonp_callback&id=1&type=object':
  case 'GET /with-post-string?a=1&b=2&callback=solidus_client_jsonp_callback&id=1&type=string':
  case 'GET /with-post-empty?a=1&b=2&callback=solidus_client_jsonp_callback':
    success();
    break;

  case 'GET /not-json':
    res.writeHead(200, {'Content-Type': 'text/plain'});
    res.end('not JSON');
    break;

  case 'GET /200-error':
    success({status: 'error', error: 'Something went wrong'});
    break;

  case 'GET /200-error-not-found':
    success({status: 'error', error: 'Item Not Found'});
    break;

  case 'GET /with-headers':
    if (req.headers.accept == 'application/json') success();
    else not_found();
    break;

  case 'GET /with-basic-auth':
    if (req.headers.authorization == 'Basic S2luZ1JvbGFuZDoxMjM0NQ==') success();
    else not_found();
    break;

  case 'GET /with-delay':
  case 'GET /api/resource.json?url=http%3A%2F%2Fsolidus.com%2Fwith-delay':
  case 'GET /with-delay?callback=solidus_client_jsonp_callback':
    setTimeout(success, 100);
    break;

  case 'GET /with-gzip':
    if (req.headers['accept-encoding'].indexOf('gzip') >= 0) {
      zlib.gzip('{"url": "/with-gzip"}', function(err, response) {
        res.writeHead(200, {'Content-Encoding': 'gzip'});
        res.end(response);
      });
    } else not_found();
    break;

  case 'POST /with-post-object?a=1&b=2':
    var body = '';
    req.on('data', function(data) {body += data;});
    req.on('end', function() {
      if (body == '{"id":1,"type":"object"}') success();
      else not_found();
    });
    break;

  case 'POST /with-post-string?a=1&b=2':
    var body = '';
    req.on('data', function(data) {body += data;});
    req.on('end', function() {
      if (body == 'id=1&type=string') success();
      else not_found();
    });
    break;

  case 'POST /with-post-empty?a=1&b=2':
    var body = '';
    req.on('data', function(data) {body += data;});
    req.on('end', function() {
      if (!body) success();
      else not_found();
    });
    break;

  default:
    not_found();
  }
};


index.js
var _ = require('underscore');
var handlebars_helper = require('handlebars-helper');
var EventEmitter = require('events').EventEmitter;
var util = require('util');

var Resource = require('./lib/resource.js');
var View = require('./lib/view.js');

var SolidusClient = function(options) {
  if (!(this instanceof SolidusClient)) return new SolidusClient(options);

  options || (options = {});
  this.resources_options = options.resources_options;
  this.context           = options.context || {};
};

// HACK: needed by universe.js...
util.inherits(SolidusClient, EventEmitter);

SolidusClient.prototype.getResource = function(options, params, callback) {
  var resource = new Resource(options, this.resources_options, params);
  resource.get(function(err, response) {
    callback(err, response ? response.data : null);
  });
};

SolidusClient.prototype.getResources = function(resources, params, callback) {
  var self = this;
  var errors = null;
  var result = {};
  var remaining = _.size(resources);

  if (!remaining) return callback(errors, result);

  _.each(resources, function(options, name) {
    self.getResource(options, params, function(err, data) {
      if (err) {
        errors = errors || {};
        errors[name] = err;
      } else {
        result[name] = data;
      }
      if (!--remaining) callback(errors, result);
    });
  });
};

SolidusClient.prototype.renderTemplate = function(template, context, template_options, callback) {
  var html, err;

  template_options = template_options || {};
  template_options.helpers = _.extend(handlebars_helper.helpers, template_options.helpers || {});

  try {
    html = template(context, template_options);
  } catch (error) {
    err = error;
  }

  callback(err, html);
};

SolidusClient.prototype.view = function(view) {
  var self = this;

  view.render = function() {
    var args = Array.prototype.slice.call(arguments);
    args.unshift(this);
    return self.render.apply(self, args);
  };

  return view;
};

SolidusClient.prototype.render = function(view, params, callback) {
  var renderer = new View(this, view);
  if (!callback && _.isFunction(params)) {
    callback = params;
    params   = null;
  }
  if (params) {
    renderer.params(params);
  }
  return callback ? renderer.end(callback) : renderer;
};

module.exports = SolidusClient;


gulpfile.js
var gulp = require('gulp');
var gutil = require('gulp-util');
var mocha = require('gulp-mocha');
var browserify = require('gulp-browserify');
var derequire = require('gulp-derequire');
var rename = require('gulp-rename');
var runSequence = require('run-sequence');
var http = require('http');
var BrowserStackTunnel = require('browserstacktunnel-wrapper');

var config = require('./test/config');

gulp.task('default', ['build']);

gulp.task('build', function() {
  return gulp
    .src('./index.js')
    .pipe(browserify({standalone: 'SolidusClient'}))
    .pipe(derequire())
    .pipe(rename('solidus-client.js'))
    .pipe(gulp.dest('./build'));
});

gulp.task('test', function(callback) {
  runSequence(
    ['build-browser-test', 'start-test-server', 'start-browserstack-tunnel'],
    'run-node-test',
    'run-browser-test',
    function() {
      runSequence(
        ['stop-test-server', 'stop-browserstack-tunnel'],
        callback);
    }
  );
});

gulp.task('node-test', function(callback) {
  runSequence(
    'start-test-server',
    'run-node-test',
    function() {
      runSequence(
        'stop-test-server',
        callback);
    }
  );
});

gulp.task('browser-test', function(callback) {
  runSequence(
    ['build-browser-test', 'start-test-server', 'start-browserstack-tunnel'],
    'run-browser-test',
    function() {
      runSequence(
        ['stop-test-server', 'stop-browserstack-tunnel'],
        callback);
    }
  );
});

gulp.task('test-server', function(callback) {
  runSequence(
    ['build-browser-test', 'start-test-server'],
    callback);
});

// TASKS

gulp.task('build-browser-test', function() {
  return gulp
    .src('./test/test.js')
    .pipe(browserify())
    .pipe(gulp.dest('./test/browser'));
});

var test_server;
gulp.task('start-test-server', function(callback) {
  test_server = http
    .createServer(config.routes)
    .listen(config.port, function() {
      gutil.log('Test server started on', gutil.colors.green(config.host));
      gutil.log('Run manual tests on', gutil.colors.green(config.host + '/test/browser/test.html'));
      callback();
    });
});

gulp.task('stop-test-server', function(callback) {
  test_server.close(callback);
});

var browserstack_tunnel;
gulp.task('start-browserstack-tunnel', function(callback) {
  browserstack_tunnel = new BrowserStackTunnel({
    key: 'TODO',
    v: true
  });

  browserstack_tunnel.start(function(err) {
    if (!err) gutil.log('BrowserStack tunnel started');
    callback(err);
  });
});

gulp.task('stop-browserstack-tunnel', function(callback) {
  browserstack_tunnel.stop(callback);
});

gulp.task('run-node-test', function(callback) {
  gulp
    .src('./test/test.js', {read: false})
    .pipe(mocha())
    .on('end', callback)
    .on('error', function() {});
});

gulp.task('run-browser-test', function(callback) {
  gulp
    .src('./test/browser-test.js', {read: false})
    .pipe(mocha({timeout: 60000}))
    .on('end', callback)
    .on('error', function() {});
});


lib/view.js
var _ = require('underscore');

var util = require('./util');

var View = function(solidus_client, view) {
  this.solidus_client = solidus_client;

  if (_.isFunction(view)) {
    this.template = view;
  } else {
    this.resources = view.resources;
    this.resources_params = view.params;
    this.preprocessor = view.preprocessor;
    this.template = view.template;
    this.template_options = view.template_options;
  }

  this.resources = this.resources || {};
  this.resources_params = this.resources_params || {};
  this.template_options = this.template_options || {};
};

View.prototype.params = function(params) {
  this.resources_params = params;
  return this;
};

View.prototype.templateOptions = function(template_options) {
  this.template_options = template_options;
  return this;
};

View.prototype.get = function(resources) {
  this.resources = resources;
  return this;
};

View.prototype.then = function(preprocessor) {
  this.preprocessor = preprocessor;
  return this;
};

View.prototype.end = function(callback) {
  var self = this;

  getResources.call(self, function(err, resources) {
    if (err) return callback(err);

    addResourcesToContext.call(self, resources);

    preprocessContext.call(self, function(err) {
      if (err) return callback(err);

      self.solidus_client.renderTemplate(self.template, self.solidus_client.context, self.template_options, callback);
    });
  });
};

// Hacks for easier testing
View.redirect = function(url) {
  window.location.href = url;
};
View.preprocessor_timeout = 10000;

// PRIVATE

var getResources = function(callback) {
  var self = this;
  self.solidus_client.getResources(self.resources, self.resources_params, function(err, resources) {
    if (err) {
      err = _.find(err, function(err, resource) {
        return !self.resources[resource].optional
      });
    }
    callback(err, resources);
  });
};

var addResourcesToContext = function(resources) {
  this.solidus_client.context.resources || (this.solidus_client.context.resources = {});
  for (var name in resources) {
    this.solidus_client.context.resources[name] = resources[name];
  }
};

var preprocessContext = function(callback) {
  var self = this;
  if (!self.preprocessor) return callback();

  var preprocessor_type = self.preprocessor.length === 1 ? preprocessContextSync : preprocessContextAsync;

  preprocessor_type.call(self, function(err, context) {
    if (err) {
      callback(err);
    } else if (_.isString(context)) {
      if (util.isNode) return callback(new Error('Cannot redirect'));
      View.redirect(context);
    } else if (_.isArray(context) && context.length === 2 && _.isNumber(context[0]) && _.isString(context[1])) {
      if (util.isNode) return callback(new Error('Cannot redirect'));
      View.redirect(context[1]);
    } else if (!_.isObject(context)) {
      callback(new Error('Preprocessor returned invalid context'));
    } else {
      self.solidus_client.context = context;
      callback();
    }
  });
};

var preprocessContextSync = function(callback) {
  var context, err;

  try {
    context = this.preprocessor(this.solidus_client.context);
  } catch (error) {
    err = error;
  }

  callback(err, context);
};

var preprocessContextAsync = function(callback) {
  var done = _.once(callback);

  setTimeout(function() {
    done(new Error('Preprocessor timeout'));
  }, View.preprocessor_timeout);

  try {
    this.preprocessor(this.solidus_client.context, function(context) {
      // In case the preprocessor is not async, we don't want to catch the callback exceptions
      setTimeout(function() {
        done(null, context);
      }, 0);
    });
  } catch (err) {
    done(err);
  }
};

module.exports = View;


lib/resource.js
var DYNAMIC_SEGMENTS = /\{([^\}]*)\}/ig;
var DEFAULT_ENCODING = 'UTF8';
var DEFAULT_SOLIDUS_API_ROUTE = '/api/';
var DEFAULT_TIMEOUT = 20000;
var NOT_FOUND = /(404|not[ -_]?found|not[ -_]?find|no[ -_]?\w+?[ -_]?found|invalid[ -_]?\w+|unknown[ -_]?\w+|do(es)?[ -_]?not[ -_]?exist|empty)/i;

var _ = require('underscore');
var qs = require('qs');
var superagent = require('superagent');
var extend = require('extend');

var util = require('./util');

// SuperAgent uses btoa for auth encoding on the client, which is missing from <IE10
// https://github.com/visionmedia/superagent#supported-browsers
if (!util.isNode && !window.btoa) window.btoa = require('./base64').encode;

var Resource = function(options, resources_options, params) {
  if (_.isString(options)) {
    initializeUrl.call(this, options, params || {});
    initializeOptions.call(this, {}, resources_options || {}, params || {});
  } else {
    initializeUrl.call(this, options.url, params || {});
    initializeOptions.call(this, _.omit(options, 'url'), resources_options || {}, params || {});
  }
};

Resource.prototype.requestType = function() {
  if (!util.isNode) {
    if (this.options.proxy) return 'proxy';
    if (this.options.jsonp || (this.options.with_credentials && !util.supportsCORS)) return 'jsonp';
  }
  return 'client';
};

Resource.prototype.requestUrl = function() {
  if (!this.url) return;

  switch (this.requestType()) {
  case 'proxy':
    return proxyRequestUrl.call(this);
  case 'jsonp':
    return jsonpRequestUrl.call(this);
  case 'client':
    return clientRequestUrl.call(this);
  }
};

Resource.prototype.get = function(callback) {
  request.call(this, 'get', null, callback);
};

Resource.prototype.post = function(data, callback) {
  request.call(this, 'post', data, callback);
};

// PRIVATE

var initializeUrl = function(url, params) {
  if (validUrl(url)) {
    var expanded = expandVariables(url, params);
    this.url             = expanded[0];
    this.dynamic         = expanded[1];
    this.expanded_params = expanded[2];
  } else {
    this.url             = null;
    this.dynamic         = false;
    this.expanded_params = [];
  }
};

var initializeOptions = function(options, resources_options, params) {
  var self = this;
  self.options = {timeout: DEFAULT_TIMEOUT};

  // Find all the matching global options
  _.each(resources_options, function(options, match) {
    var matcher = new RegExp(match, 'ig');
    if (matcher.test(self.url)) extend(true, self.options, options);
  });

  // Merge the resource options
  extend(true, self.options, options);

  // Expand the query options
  _.each(self.options.query, function(value, name) {
    if (!_.isString(value)) return;
    var expanded = expandVariables(value, params);
    self.options.query[name] = expanded[0];
    if (expanded[1]) self.dynamic = true;
    self.expanded_params = self.expanded_params.concat(expanded[2]);
  });
};

var validUrl = function(url) {
  return /^https?:\/\/.+/.test(url);
};

var expandVariables = function(string, params) {
  var dynamic = false;
  var expanded = [];
  var result = string.replace(DYNAMIC_SEGMENTS, function(match, variable) {
    dynamic = true;
    if (params[variable]) expanded.push(variable);
    return params[variable] || '';
  });
  return [result, dynamic, expanded];
};

var request = function(method, data, callback) {
  var self = this;

  if (!this.url) return callback(null, {});
  if (this.options.required_params) {
    var missing_params = this.options.required_params.filter(function (param) {return self.expanded_params.indexOf(param) === -1});
    if (missing_params.length) return callback(new Error('Missing required params: ' + missing_params.join(', ')), {});
  }

  var request_type;
  switch (this.requestType()) {
  case 'proxy':
    request_type = proxyRequest;
    break;
  case 'jsonp':
    request_type = jsonpRequest;
    break;
  case 'client':
    request_type = clientRequest;
    break;
  }

  var result = {request_time: new Date().getTime()};
  request_type.call(this, method, data, function(err, res, data) {
    result.response_time = new Date().getTime();
    result.response = res;
    result.data = data;
    callback(err, result);
  });
};

var proxyRequest = function(method, data, callback) {
  if (method != 'get') return callback('Invalid proxy method: ' + method);

  var request = superagent.get(proxyRequestUrl.call(this));

  if (this.options.timeout) request.timeout(this.options.timeout);

  superAgentRequest(request, callback);
};

var proxyRequestUrl = function() {
  var route = (this.options.solidus_api_route || DEFAULT_SOLIDUS_API_ROUTE) + 'resource.json';
  var url   = buildUrl.call(this, this.url);
  return route + '?url=' + encodeURIComponent(url);
};

var jsonpRequest = function(method, data, callback) {
  if (method != 'get' & method != 'post') return callback('Invalid JSONP method: ' + method);

  var self = this;
  var callbackName = 'solidus_client_jsonp_callback_' + Math.round(100000 * Math.random());
  var url = buildUrl.call(self, jsonpRequestUrl.call(self), {callback: callbackName}, _.isObject(data) ? data : {});
  if (_.isString(data)) url += '&' + data;

  var script = document.createElement('script');
  script.src = url;

  var timer;
  var first = true;

  window[callbackName] = function(data) {
    var timedout = data instanceof Error;

    // Cancel timeout
    if (timer) {
      clearTimeout(timer);
      timer = null;
    }

    // Do not cleanup if there was a timeout, in case the real request ever comes back
    if (!timedout) {
      document.getElementsByTagName('head')[0].removeChild(script);
      delete window[callbackName];
    }

    // Only run user callback once
    if (first) {
      first = false;
      timedout ? callback(data) : processResponse({status: 200}, data, callback);
    }
  };

  if (self.options.timeout) {
    timer = setTimeout(function() {
      timer = null;
      var err = new Error('Timeout of ' + self.options.timeout + 'ms exceeded');
      err.callbackName = callbackName; // For testing purposes
      window[callbackName](err);
    }, self.options.timeout);
  }

  document.getElementsByTagName('head')[0].appendChild(script);
};

var jsonpRequestUrl = function(data) {
  return buildUrl.call(this, this.url);
};

var clientRequest = function(method, data, callback) {
  if (method != 'get' & method != 'post') return callback('Invalid method: ' + method);

  var url     = clientRequestUrl.call(this);
  var request = method == 'get' ? superagent.get(url) : superagent.post(url).send(data);

  if (this.options.headers) request.set(this.options.headers);
  if (this.options.auth) request.auth(this.options.auth.user, this.options.auth.pass);
  if (this.options.with_credentials && !util.isNode) request.withCredentials();
  if (this.options.timeout) request.timeout(this.options.timeout);

  superAgentRequest(request, callback);
};

var clientRequestUrl = function() {
  return buildUrl.call(this, this.url);
};

var buildUrl = function(url) {
  var start = url.indexOf('?') + 1;
  var query = [];
  if (start > 0) query.push(qs.parse(url.substring(start)));
  if (this.options.query) query.push(this.options.query);
  query = query.concat(Array.prototype.slice.call(arguments, 1));
  query = _.extend.apply(_, query);

  if (this.options.query_options && this.options.query_options['objectFormat'] === 'json') {
    _.each(query, function(value, key) {
      if (_.isObject(value) && !_.isArray(value)) query[key] = JSON.stringify(value);
    });
  }

  query = qs.stringify(query, this.options.query_options);
  if (query) {
    return (start > 0 ? url.substring(0, start - 1) : url) + '?' + query;
  } else {
    return url;
  }
};

var superAgentRequest = function(request, callback) {
  // Force buffering of response bodies as res.text
  // http://visionmedia.github.io/superagent/#buffering-responses
  if (util.isNode) request.buffer();

  // Not all weird browser errors are catched by SuperAgent
  try {
    request.end(function(err, res) {
      // In case the request is not async, we don't want to catch the callback exceptions
      setTimeout(function() {
        // SuperAgent considers statuses <200 and >=300 as errors
        // http://visionmedia.github.io/superagent/#error-handling
        if (err) return callback(err, res);
        processResponse(res, null, callback);
      }, 0);
    });
  } catch (err) {
    callback(err);
  }
}

var processResponse = function(res, data, callback) {
  processData(res, data, function(err, data) {
    if (err) return callback(err, res);
    processDataStatus(data, function(err) {
      callback(err, res, data);
    });
  });
};

var processData = function(res, data, callback) {
  var err;
  if (!_.isObject(data)) {
    try {
      data = JSON.parse(res.text.toString(DEFAULT_ENCODING));
    } catch (error) {
      err = new Error('Invalid JSON');
      err.error = error.message;
    }
  }
  callback(err, data);
};

var processDataStatus = function(data, callback) {
  var err;
  if (String(data.status).toLowerCase() == 'error') {
    var message = (data.messages || []).concat([data.error, data.message, data.response, data.body]).join(' ');
    err = new Error('Invalid data status');
    err.status = message.match(NOT_FOUND) ? 404 : 500;
  }
  callback(err);
};

module.exports = Resource;
